pkgs/http_client_conformance_tests/lib/src/abort_tests.dart

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2025, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:async/async.dart';
+import 'package:http/http.dart';
+import 'package:stream_channel/stream_channel.dart';
+import 'package:test/test.dart';
+
+import 'abort_server_vm.dart'
+    if (dart.library.js_interop) 'abort_server_web.dart';
+
+/// Tests that the client supports aborting requests.
+///
+/// If [supportsAbort] is `false` then tests that assume that requests can be
+/// aborted will be skipped.
+///
+/// If [canStreamResponseBody] is `false` then tests that assume that the
+/// [Client] supports receiving HTTP responses with unbounded body sizes will
+/// be skipped.
+///
+/// If [canStreamRequestBody] is `false` then tests that assume that the
+/// [Client] supports sending HTTP requests with unbounded body sizes will be
+/// skipped.
+void testAbort(
+  Client client, {
+  bool supportsAbort = true,
+  bool canStreamRequestBody = true,
+  bool canStreamResponseBody = true,
+}) {
+  group('abort', () {
+    late String host;
+    late StreamChannel<Object?> httpServerChannel;
+    late StreamQueue<Object?> httpServerQueue;
+    late Uri serverUrl;
+
+    setUp(() async {
+      httpServerChannel = await startServer();
+      httpServerQueue = StreamQueue(httpServerChannel.stream);
+      host = 'localhost:${await httpServerQueue.nextAsInt}';
+      serverUrl = Uri.http(host, '');
+    });
+    tearDownAll(() => httpServerChannel.sink.add(null));
+
+    test('before request', () async {
+      final abortTrigger = Completer<void>();
+      final request = AbortableRequest(
+        'GET',
+        serverUrl,
+        abortTrigger: abortTrigger.future,
+      );
+      abortTrigger.complete();
+
+      expect(
+        client.send(request),
+        throwsA(isA<AbortedRequest>()),
+      );
+    });
+
+    test('during request stream', () async {
+      final abortTrigger = Completer<void>();
+
+      final request = AbortableStreamedRequest(
+        'POST',
+        serverUrl,
+        abortTrigger: abortTrigger.future,
+      );
+
+      final response = client.send(request);
+      request.sink.add('Hello World'.codeUnits);
+
+      abortTrigger.complete();
```

Comment:
Closing the sink will break if we change the code to:
```
      expect(
        response,
        throwsA(isA<AbortedRequest>()),
      );

      for (int i = 0; i < 10000; ++i) {
        request.sink.add('Hello World'.codeUnits);
        await Future.delayed(const Duration());
      }
```

```
  Bad state: Cannot add event after closing
```

Reply:
So my suggestion would be to either:
 - make this a forbidden action (the user shouldn't be trying to pipe in after aborting)
 - make a dummy consumer which silently switches away from the aborted request and drops incoming data

Reply:
The docs for `abort` say:

  /// Using the [IOSink] methods (e.g., [write] and [add]) has no effect after
  /// the request has been aborted

Reply:
Not at my computer right now so can't investigate, but does addStream use that? I suspect it might use a more efficient method on the stream consumer (but I might be wrong).

Reply:
Maybe my test should just considered malformed for not closing the sink.

Reply:
I think that I changed my mind about this. Since streaming uploads are not supported by `BrowserClient`, it will never cancel the stream subscription. So I don't think that we need to do so in `IOClient`.